### PR TITLE
BUG: Fix MLE for Nakagami `nakagami_gen.fit`

### LINKED ISSUE #10908: Nakagami fitting doesn't converge  (scipy.stats)
Hi guys, 

I'm trying use scipy.stats to analyses communication channels, basically I'm calculating "rayleigh" and "nakagami" distributions to some RSSI measurements.
The most cases both distribution fit converge, but some times an issue is occurring to distribution fit for nakagami, the shape (or "mu") return a strange value.

The example below I have 5000  samples and the nakagami distribution return mu=17.63, I set this same data in matlab and return mu=4.
Measurements
https://imgur.com/l1eYgsP

nakagami pdf
https://imgur.com/RZJIweT

Anyone have an ideia about what happen?

thanks

### PR DESCRIPTION
**Pull request by jjerphan**
In some configurations (see scipy#10908), the values for MLE are completely out for Nakagami.

#### Reference issue

Closes gh-10908

#### What does this implement/fix?

It add a better first estimates of the parameters as well as tests for `nakagami_gen.fit`.

#### Additional information

Alternatively, [the MLE estimators have been analytically derived for Nakagami](https://ieeexplore.ieee.org/document/913153). 
One could implement them in `nakagami_gen.fit` directly.

Also, maybe other distributions should be checked for their MLE using parametrised tests.

If this is relevant, I can work on this to adapt the tests that I've done here to others (`rv_continuous`) distributions.

### CODE REVIEW

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5301]: If you parametrize w.r.t. each variable separately, e.g.:
```
@pytest.mark.parametrize('nu', ...)
@pytest.mark.parametrize('loc', ...)
...
```
I believe you get the same behavior without `itertools`.

That said, I'm not sure if we would want so many combinations to run every time a push is made to a PR, as we have limited CI resources to work with. Instead of all these, can you pick one or a few cases didn't work before your change and work after?

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5807]:
```suggestion
        return loc, scale
```
`_fitstart` needs to return an estimate of the shape parameter(s), too.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5804]: Is there a reference for this, or by "empirically" do you mean in informal experiments?

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5805]: Please double-check, but I think `data` might not be a NumPy array at this point; e.g. it could be a list. If that's right, consider changing this to `np.mean(data)`

**Review by mdhaber — APPROVED** [review]
Thanks for the suggestion @jjerphan. I think this would be welcome if we can verify that it is a sound approach and fixes more cases than it breaks.

Please do [run tests locally](http://scipy.github.io/devdocs/dev/contributor/runtests.html#runtests) before pushing, though, to conserve CI resources.

**Review by mdhaber — CHANGES_REQUESTED** [review]
Oops I accidentally clicked "Approve" instead of "Request Changes"

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py]: You are right, I addressed your two points in 6a10aff0370833b176f14e34944a3735a1f65f6c.

File: scipy/stats/_continuous_distns.py
Comment (jjerphan) [reply scipy/stats/_continuous_distns.py]: I forgot to commit all my changes before pushing at first. This is addressed in 1a0164d1c793d1670873f21472b811625cf7e04f.

File: scipy/stats/_continuous_distns.py
Comment (jjerphan) [reply scipy/stats/_continuous_distns.py]: You are right. I've included a reference to the original issue in 8dd8d55668899160bab065ebc89e1d9e2a2a4638.

File: scipy/stats/_continuous_distns.py
Comment (jjerphan) [reply scipy/stats/_continuous_distns.py]: Yes, we better use your proposition for safety. This is addressed in 7ba2e55ed61663cd5d6b1626c247439451e05253.

**Comment by jjerphan** [comment]
Thanks for the review! I've responded in your threads.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:4]:
```suggestion
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5304]: Are the rest arbitrary, or do they currently fail on master?

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5307]: The tolerance is very loose here. Let me think about this change some more before I suggest a test.
(BTW, the documentation for assert_almost_equal now mentioned that assert_allclose is preferred.)

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5303]: 21 or those 27 combinations fail on master.

From the experimentations I've done with data generation, a great number of cases (easily more than half of them) show that the computations of the ML estimators for Nakagami is not stable. This motivated my original and more exhaustive parametrisation of this test.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5303]: Gotcha. Thanks!

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5307]: Yes, that's because I use a small value for `size`. As said in the PR description, maybe implementing analytical ML estimators for Nakagami would be preferred. I am reading about other issues and PRs about scipy.stats to know what the current direction is regarding MLE, especially analytical solutions or not.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5307]: Yes, analytical fits are preferred, but sometimes they are a lot more work.
I am going to guess that the analytical fit for the location is actually the minimum of the data rather than the mean, btw. That's what I'm going to look into.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5307]: You can take a look at some of the other distributions with recently overridden `fit` methods in gh-11782 (more specifically [here](https://github.com/scipy/scipy/issues/11782#issuecomment-633188975)).

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5307]: Thanks, this is useful information to me!

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5307]: I think this PR can be seen as a hot-fix for now. I can have a look at the MLE for the general form of Nakagami in a second PR: let me know. 🙂

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py]: I think that the analytical MLE for scale, for SciPy's parameterization, is
image

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py]: Also, the mean isn't a good guess for the location, because the support of the distribution tells us that `xi - loc >= 0` for all xi, so `loc <= min(xi)`. I don't the equality provides the exact MLE, but I think `loc = np.min(xi)` and scale as provided above would be theoretically justified guesses.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5808]:
```suggestion
        # Theoretically justified estimates (see gh-10908, gh-13396)
        loc = np.min(data)
        scale = np.sqrt(np.sum((x - loc)**2) / len(x))
```
Would you check locally how this does?

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py]: Thanks for the details. Do you have a reference for those results? This way I could cite them in the implementation.

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py]: Please do provide the nasty equation so that I can see how it compares.

File: scipy/stats/_continuous_distns.py
Comment (jjerphan) [reply scipy/stats/_continuous_distns.py:5808]: It performs a bit better. Do you have a reference that I can include as a theoretical justification?

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:5808]: I can write one.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py]: OK, I'll include it in a writeup.

File: scipy/stats/_continuous_distns.py
Comment (jjerphan) [reply scipy/stats/_continuous_distns.py:5808]: Thanks for wrapping this up!
Probably we can find a proper section in Scipy's documentation for those MLE derivations.
@WarrenWeckesser: I saw that you were reviewing some others PRs for scipy.stats; as a core-dev, do you have suggestions for mathematical notes?

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:5808]: @WarrenWeckesser has often written notes like this in a PDF outside of the SciPy repo. I think you'll find them in some of the MLE PRs. The ideal place would be the [distribution tutorials](https://docs.scipy.org/doc/scipy/reference/tutorial/stats/continuous.html#continuous-distributions-in-scipy-stats), which aren't really tutorials but lists of technical information about each distribution.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5302]: I think we could make this test more sensitive by checking that the likelihood equations are satisfied (those partial derivatives are nearly zero). Do you agree?

I think it would also be a good idea to check the special case in which `'nu' = 0.5` and for which the `fit` method fixes the parameter at that value, e.g. `stats.nakagami.fit(samples, f0=0.5)`. In this case, the likelihood equation for location will _not_ be satisfied, rather, you would check that `loc = np.min(samples)`. 

I think we're getting really close here. With these changes we've discussed, I think we can merge this.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5303]: Consider mentioning here that this is a regression test for gh-13396 and that 21 of the 27 test cases failed prior to that PR.

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [reply scipy/stats/_continuous_distns.py:5808]: +1 for distribution "tutorials"

Comment (jjerphan) [reply scipy/stats/_continuous_distns.py:5808]: I'd add a section in the [associated Nakagami page](https://docs.scipy.org/doc/scipy/reference/tutorial/stats/continuous_nakagami.html) for the ML estimators derivation using [@mdhaber's notes](https://colab.research.google.com/drive/1MODsLLxzBRXy4KbWdS3T3QCKY5NDMWPc?usp=sharing).

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: > I think we could make this test more sensitive by checking that the likelihood equations are satisfied (those partial derivatives are nearly zero). Do you agree?

I do agree. Generally, such checks could be added for ML estimators of distributions without constraints on their supports.
This is just another idea for gh-11782.

> I think it would also be a good idea to check the special case in which 'nu' = 0.5

I'll add an additional test with a comment explaining the motives you gave.

File: scipy/stats/_continuous_distns.py
Comment (jjerphan) [reply scipy/stats/_continuous_distns.py]: @mdhaber: I've added your notes in the companion page of Nakagami in ff6a27429c468ab587b8f0471466590ae7b42f2a.

File: doc/source/tutorial/stats/continuous_nakagami.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_nakagami.rst:34]: This is showing up in the list of distributions:
image

Maybe because of the `====`? Maybe make it a lower-level heading. Are you able to [build docs locally](http://scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#rendering-documentation) to see if that works?

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: OK, let me know when you'd like me to take a look!

File: doc/source/tutorial/stats/continuous_nakagami.rst
Comment (jjerphan) [reply doc/source/tutorial/stats/continuous_nakagami.rst:34]: OK, `===` creates sections while I wanted to create a subsection, in that case `---` is used.

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: > In this case, the likelihood equation for location will not be satisfied, rather, you would check that `loc = np.min(samples)`.

Generally, we need to verify the constraint set on μ, which is not yet taken into account.
What if we introduce a new objective function combining the nnlf and a log barrier for the constraint on μ?

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5304]: Done in c3683c3.

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: > I'll add an additional test with a comment explaining the motives you gave.
Done in c3683c3.

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5304]: Done in 5ca7b1d2bdaf2255800ad9ea2ce7be9780fefe58.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: I think the optimization already does approximately what we want it to do here.`fit` already has a barrier of sorts... improving it is beyond the scope of this PR. (@WarrenWeckesser have been looking into that separately.)

File: doc/source/tutorial/stats/continuous_nakagami.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_nakagami.rst:48]:
```suggestion
The log-likelihood function is therefore
```
Not a big deal, but the original doesn't have colons because the equation is considered part of the sentence, and no punctuation would be appropriate if the equation were expressed as words.
Thanks for fixing the `\geq`s, though! `>=` is what I would type into code or MS Word Equation editor; sometimes I forget that I'm using LaTeX!

**Review by mdhaber — APPROVED** [review]
Looks good to me!
@ev-br the math is now in the tutorial as you suggested. Did you want to review this?

File: doc/source/tutorial/stats/continuous_nakagami.rst
Comment (jjerphan) [reply doc/source/tutorial/stats/continuous_nakagami.rst:48]: > Not a big deal, but the original doesn't have colons because the equation is considered part of the sentence, and no punctuation would be appropriate if the equation were expressed as words.

I see: I'll remove the extra-punctuation than I've added.

>  Thanks for fixing the \geqs, though! >= is what I would type into code or MS Word Equation editor; sometimes I forget that I'm using LaTeX!

No problems. 😉

Comment (mdhaber) [reply doc/source/tutorial/stats/continuous_nakagami.rst:48]: Great! We'll give others a few days to review. I'll merge on Monday or so unless there are new things that come up.

File: scipy/stats/tests/test_distributions.py
Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: On some tests the derivatives aren't all null; I am inspecting it now before pushing modifications of the test.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: You'll have to decide whether the difference is significant or not. For example, here is the expression for the partial derivative with respect to location.
image

Even if the derivative is not very small, is it small relative to each of those two terms? Do they mostly cancel each other out?

If they are _not_ canceling each other out, we need to look a little more closely. I would suggest trying to solve the system of equations so that they _are_ all zero, and then see if the log-likelihood function:
```python3
data = ...
def ll(v, u, s):
    return np.sum(np.log(nakagami.pdf(data, v, u, s)))
```
is larger with the parameters that do zero the partial derivatives than it is with the solution returned by `fit`. 

If `fit` gives us a larger log-likelihood than the solution to our system of equations, then something is wrong with our equations. If it's the other way around, then `fit` is not converging to the optimal solution.

I'm not so concerned with the latter. Numerical optimization isn't guaranteed to achieve the optimal solution. All we're trying to do in this PR is improve the guess; we still can't guarantee that the optimizer will find the optimal solution.

If the problem is our system of equations, we'll want to try to fix the error in the tutorial before we merge.

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: In the case of the partial derivative of the likelihood w.r.t μ, since there's a constraint on its support (that is μ ≤ min_i x_i), it will hardly be equal to zero (i.e. the feasible optimal value for μ is not the optimum for the log-likelihood). I can observe that on most tests. For the two others partial derivatives, they are absolutely close to 0 up to 1e-3.

As for the system resolution, I have found that the log-likelihood associated to the solutions obtained with `scipy.optimize.root` (with Levenberg-Marquardt, for e.g) is less than the one of the solution obtains with `fit` for most tests (which is good!). For the others tests, the optimization routine/scheme of `scipy.optimize.root` yields negative values for ν (as the constraints have not been specified) which make the test fail. See [this gist](https://gist.github.com/jjerphan/4bc8e025919cbfcf1eac86635f98ccb1) for reproducibility.

As you said, improving the numerical optimisation routine is out of the scope of this PR. Though I am curious to have a look at own constraints of supports for parameters are or can be handle for `rv_continuous` subclasses' MLE/`fit` generally.

---

`tl;dr`: I've augmented the test with a check on the necessary condition for optimality on two of the partial derivatives in 72fa639b43f6a103b71087bce12d9fd955978d06.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: > In the case of the partial derivative of the likelihood w.r.t μ, since there's a constraint on its support (that is μ ≤ min_i x_i), it will hardly be equal to zero (i.e. the feasible optimal value for μ is not the optimum for the log-likelihood). I can observe that on most tests.

Well, unless something else is going, either
* the partial derivative should be zero or 
* the constraint should be active (that is, μ ≤ min_i x_i). 

We see the constraint active when nu is really close to 0.5. For all the other cases, the partial derivative looks nearly zero to me:

```python3
nus= [1.6, 2.5, 3.9]
locs = [25.0, 10, 35]
scales = [13, 5, 20]

for nu, loc, scale in product(nus, locs, scales):
    N = 100
    samples = stats.nakagami.rvs(size=N, nu=nu, loc=loc,
                                  scale=scale, random_state=1337)
    v, u, s = stats.nakagami.fit(samples)

    x = samples
    dldu = (-2*v + 1) * np.sum(1/(x - u)) + 2*v/s**2 * np.sum(x-u)
    print(dldu)
```

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: > For all the other cases, the partial derivative looks nearly zero to me:

When I run this modified version of the script (with imports and printing):

```python
import numpy as np
from scipy import stats
from itertools import product
nus= [1.6, 2.5, 3.9]
locs = [25.0, 10, 35]
scales = [13, 5, 20]
for nu, loc, scale in product(nus, locs, scales):
     N = 100
     samples = stats.nakagami.rvs(size=N, nu=nu, loc=loc,
                                   scale=scale, random_state=1337)
     v, u, s = stats.nakagami.fit(samples)
 
     x = samples
     dldu = (-2*v + 1) * np.sum(1/(x - u)) + 2*v/s**2 * np.sum(x-u)
     print(f"(nu,loc,scale)=({nu}, {loc}, {scale}) \t\t dldu={dldu}")
     print(f"(v,u,s)=({v:.2f}, {u:.2f}, {s:.2f}) \t\t min(sample)={np.min(samples)}")
     print()
```
in most cases, I do not get a partial derivative which is null:
```
(nu,loc,scale)=(1.6, 25.0, 13) 		 dldu=267187821021946.56
(v,u,s)=(0.03, 26.86, 1.91) 		 min(sample)=26.862845727959257

(nu,loc,scale)=(1.6, 25.0, 5) 		 dldu=4.9108531364083774e-05
(v,u,s)=(1.74, 24.83, 5.26) 		 min(sample)=25.716479126138175

(nu,loc,scale)=(1.6, 25.0, 20) 		 dldu=136749576269998.34
(v,u,s)=(0.01, 27.87, 2.09) 		 min(sample)=27.8659165045527

(nu,loc,scale)=(1.6, 10, 13) 		 dldu=528323667636744.5
(v,u,s)=(0.03, 11.86, 2.00) 		 min(sample)=11.862845727959257

(nu,loc,scale)=(1.6, 10, 5) 		 dldu=200508511174458.75
(v,u,s)=(0.14, 10.72, 1.71) 		 min(sample)=10.716479126138175

(nu,loc,scale)=(1.6, 10, 20) 		 dldu=547515423430104.6
(v,u,s)=(0.01, 12.87, 2.14) 		 min(sample)=12.865916504552702

(nu,loc,scale)=(1.6, 35, 13) 		 dldu=134367465776150.02
(v,u,s)=(0.02, 36.86, 1.68) 		 min(sample)=36.862845727959254

(nu,loc,scale)=(1.6, 35, 5) 		 dldu=96135730775175.7
(v,u,s)=(0.16, 35.72, 1.84) 		 min(sample)=35.716479126138175

(nu,loc,scale)=(1.6, 35, 20) 		 dldu=137355339523805.89
(v,u,s)=(0.01, 37.87, 1.94) 		 min(sample)=37.8659165045527

(nu,loc,scale)=(2.5, 25.0, 13) 		 dldu=261796801631422.75
(v,u,s)=(0.03, 28.33, 1.69) 		 min(sample)=28.32581019355903

(nu,loc,scale)=(2.5, 25.0, 5) 		 dldu=194977126600637.5
(v,u,s)=(0.15, 26.28, 1.50) 		 min(sample)=26.27915776675347

(nu,loc,scale)=(2.5, 25.0, 20) 		 dldu=275506932660216.2
(v,u,s)=(0.01, 30.12, 1.67) 		 min(sample)=30.11663106701389

(nu,loc,scale)=(2.5, 10, 13) 		 dldu=504184655175108.25
(v,u,s)=(0.05, 13.33, 2.16) 		 min(sample)=13.325810193559029

(nu,loc,scale)=(2.5, 10, 5) 		 dldu=283985415051308.0
(v,u,s)=(0.25, 11.28, 1.83) 		 min(sample)=11.279157766753473

(nu,loc,scale)=(2.5, 10, 20) 		 dldu=8.439969640505751e-06
(v,u,s)=(2.77, 9.05, 21.20) 		 min(sample)=15.11663106701389

(nu,loc,scale)=(2.5, 35, 13) 		 dldu=134522106712885.08
(v,u,s)=(0.02, 38.33, 1.51) 		 min(sample)=38.32581019355903

(nu,loc,scale)=(2.5, 35, 5) 		 dldu=0.0002123006828327334
(v,u,s)=(2.77, 34.76, 5.30) 		 min(sample)=36.27915776675347

(nu,loc,scale)=(2.5, 35, 20) 		 dldu=136948858351787.02
(v,u,s)=(0.01, 40.12, 1.80) 		 min(sample)=40.11663106701389

(nu,loc,scale)=(3.9, 25.0, 13) 		 dldu=264856304242045.47
(v,u,s)=(0.03, 29.85, 1.51) 		 min(sample)=29.847553943506007

(nu,loc,scale)=(3.9, 25.0, 5) 		 dldu=134996579286915.28
(v,u,s)=(0.26, 26.86, 1.54) 		 min(sample)=26.864443824425386

(nu,loc,scale)=(3.9, 25.0, 20) 		 dldu=135290418754594.25
(v,u,s)=(0.02, 32.46, 1.83) 		 min(sample)=32.45777529770155

(nu,loc,scale)=(3.9, 10, 13) 		 dldu=-1.2802542002532391e-05
(v,u,s)=(4.21, 9.53, 13.61) 		 min(sample)=14.847553943506009

(nu,loc,scale)=(3.9, 10, 5) 		 dldu=339393765797140.2
(v,u,s)=(0.20, 11.86, 1.43) 		 min(sample)=11.864443824425388

(nu,loc,scale)=(3.9, 10, 20) 		 dldu=269159753937960.53
(v,u,s)=(0.02, 17.46, 2.11) 		 min(sample)=17.457775297701552

(nu,loc,scale)=(3.9, 35, 13) 		 dldu=127868092068877.98
(v,u,s)=(0.05, 39.85, 1.84) 		 min(sample)=39.84755394350601

(nu,loc,scale)=(3.9, 35, 5) 		 dldu=75980042179752.31
(v,u,s)=(0.23, 36.86, 1.61) 		 min(sample)=36.864443824425386

(nu,loc,scale)=(3.9, 35, 20) 		 dldu=134193438617571.53
(v,u,s)=(0.02, 42.46, 1.84) 		 min(sample)=42.45777529770155

```

> the constraint should be active (that is, μ ≤ min_i x_i).

I think that it is rather μ = min_i x_i if the constraint is active; see 4.1.1 (p. 127) of [Stephen Boyd and Lieven Vandenberghe. 2004. Convex Optimization. Cambridge University Press, USA.](https://web.stanford.edu/~boyd/cvxbook/bv_cvxbook.pdf) for reference.

The (quasi-)equality is also something that we tend to observe when the partial derivative is not null.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: Weird... I run that script and get all nearly zero! For example, the first few outputs:
```
(nu,loc,scale)=(1.6, 25.0, 13) 		 dldu=-5.5426409719672165e-05
(v,u,s)=(1.74, 24.55, 13.66) 		 min(sample)=26.862845727959257

(nu,loc,scale)=(1.6, 25.0, 5) 		 dldu=-0.00036703005109472997
(v,u,s)=(1.74, 24.83, 5.26) 		 min(sample)=25.716479126138175

(nu,loc,scale)=(1.6, 25.0, 20) 		 dldu=-1.5228397325373066e-05
(v,u,s)=(1.74, 24.30, 21.02) 		 min(sample)=27.8659165045527

(nu,loc,scale)=(1.6, 10, 13) 		 dldu=-1.1207026602733094e-05
(v,u,s)=(1.74, 9.55, 13.66) 		 min(sample)=11.862845727959257
```

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: > I think that it is rather μ = min_i x_i if the constraint is active

Yes, I copy-pasted and forgot to change it to an equality sign, which is what active means.

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: Hum, this is weird. Here are the versions that I use:
 - Python 3.9.1
 - scipy: '1.7.0.dev0+04a36f7' (this fork)
 - numpy: '1.19.4'

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: My bad: I just was on another branch when executing, your results are right.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5302]: Now in case you'd like to see a case where the optimizer _doesn't_ converge to the right answer, change `random_state` to 1...

Comment (jjerphan) [reply scipy/stats/tests/test_distributions.py:5302]: I guess 3b467ef5362e806669b9a61da6390108a35fbbff closes this thread. 😁

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5314]:
```suggestion
            return ((-2*nu + 1) * np.sum(1/(samples - loc))
```
Not worth re-running the test suite for this change, but implicit line breaks (implied by the fact that there is an open left paranthesis) are preferred according to PEP8.

**Comment by mdhaber** [comment]
@jjerphan in see my review was re-requested, but i already approved. Since I was very closely involved in the development of the math and documentation, I think I should let someone else review and merge, though. We'll certainly make sure this gets in the next release, though. Sound good?

**Comment by jjerphan** [comment]
> @jjerphan in see my review was re-requested, but i already approved. Since I was very closely involved in the development of the math and documentation, I think I should let someone else review and merge, though. We'll certainly make sure this gets in the next release, though. Sound good?

Yes, sure. 🙂

**Comment by jjerphan** [comment]
[A CI test](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=10249&view=logs&j=e177c0b2-361c-5cee-6783-4418e98d3ffa&t=663e3bfe-bbaf-5419-7ee8-d9a433818df8&l=638) is now failing because the method of moments for `fit` has been merged (see #11695) and because the proposed first estimates (derived for the MLE) in this PR aren't suitable for this method.

How should we proceed since `_fitstart` is used in both method? Should we just use the previous estimates for the method of moments? This would make the failing test pass, but there might be better estimates.

This problem might appear for other PRs treating #11782.

**Comment by mdhaber** [comment]
> A CI test is now failing...

I think the most appropriate thing to do in this PR - as we don't want to increase the scope when this is all ready to go - is add `nakagami` to the list of failing fits [here](https://github.com/scipy/scipy/blob/f304bd01bf3d70a4337f616bc9c2352c76f3cd7c/scipy/stats/tests/test_fit.py#L39). I think that will skip the test, and I think that's fine. Please test locally before pushing to make sure it actually skips. 

Adding separate guesses for method of moments is not done for any other distributions yet, so I don't think we should start in this PR. Also, I don't think we would want to revert to the original initializer for method of moments unless we have reason to believe - other than this single example - that it typically provides a better guess than this PR does. Are there other examples? 

When we merge this PR, we can open up an issue about it.

**Comment by jjerphan** [comment]
> > A CI test is now failing...
> 
> I think the most appropriate thing to do in this PR - as we don't want to increase the scope when this is all ready to go - is add `nakagami` to the list of failing fits [here](https://github.com/scipy/scipy/blob/f304bd01bf3d70a4337f616bc9c2352c76f3cd7c/scipy/stats/tests/test_fit.py#L39). I think that will skip the test, and I think that's fine. Please test locally before pushing to make sure it actually skips.

I'll do it now.

> Adding separate guesses for method of moments is not done for any other distributions yet, so I don't think we should start in this PR. Also, I don't think we would want to revert to the original initializer for method of moments unless we have reason to believe - other than this single example - that it typically provides a better guess than this PR does. Are there other examples?

I agree. I do not have any other examples.

**Comment by jjerphan** [comment]
Hi @ev-br,
Are you interested in reviewing this PR? 🙂

**Comment by mdhaber** [comment]
I'll go ahead and merge this. Thanks @jjerphan! Looking forward to the next one.

**Comment by jjerphan** [comment]
Thanks for your guidance @mdhaber!

**Pull request merged**
